docs/pages/machine-id/architecture.mdx

```diff
@@ -0,0 +1,154 @@
+authority.
+
+## Role Impersonation
+
+At the core of Machine ID is the concept of role impersonation.
+
+Role Impersonation allows a user to generate credentials with a set of requested
+roles. The user does not have to hold these roles, but must have been granted 
+permission to impersonate them. The impersonated credentials still include
+the username of the user that generated them, so actions can be attributed to 
+the user.
+
+These credentials can then be used to complete any action that is allowed by the
+role's configured permissions. 
+
+In the case of Machine ID, the bot user is assigned a bot role, which includes 
+permissions to impersonate the roles that the user has configured.
+
+## Onboarding
+
+Onboarding is the process by which `tbot` initially authenticates with the 
+Teleport auth service in order to receive credentials.
+
+Machine ID leverages the existing Token resource within Teleport, with the
+token containing an additional `botName` field that identifies the bot user
+associated with the token. 
+
+Machine ID currently supports two methods of joining that have some key
+differences.
+
+Static Token:
+
+- The name of the token is used as an opaque secret needed to join the Teleport 
+  cluster. This means it must be stored and communicated securely.
+- Once used, the token resource self-destructs. This means it can only be used 
+  to join a single bot to a Teleport cluster.
+- The certificates exchanged for the token are 
+  [renewable](#renewable-certificates).
+
+
+Dynamic join tokens (e.g AWS IAM):
+
+- These tokens rely on some kind of external system that allows the bot to
+  prove it is allowed to join the cluster. The name of the token identifies the
+  Token resource in Teleport that contains the configuration.
+- The token can be used to join as many bots as you want, and do not self
+  destruct in the same way that static tokens do.
+- The certificates exchanged for the token are not renewable. When the bot wants
+  to renew its certificates, it simply repeats the original join steps.
+
+Where possible, you should prefer to use a dynamic join token over a static
+token as this eliminates the need to handle a secret.
+
+## Renewable certificates
+
+Human Teleport users are not able to renew their credentials without
+re-authenticating. As Machine ID must work without manual human intervention,
+bots that have joined using static tokens are granted credentials that are
+marked as renewable. This means that the bot can use these credentials to
+fetch a new set of credentials with an expiry date further in the future.
+
+In order to mitigate the risk of bot user credentials being stolen, and then
+continually renewed by a malicious actor, renewable bot user certificates
+include a **generation counter**.
+
+The generation counter is stored against the user in the database and within the
+certificate. This counter is incremented each time the user renews their 
```

Comment:
Does this counter truly increment or does it follow some other kind of random generation? What is preventing success of a brute force style attack based on a smaller subset of incremented values?

Reply:
Yes, this will be a monotonically increasing value. I'm not quite sure I understand the concerns re: brute force/dictionary style attack, this field is encoded within the certificate by the auth service and isn't a value provided by a client.

Reply:
My mistake, I misunderstood, i see how the ephemeral token and user permission work together to lock things down if any attack is made, I must have admittedly been reading too quickly. Marking this as resolved and adding my approval, this describes something relatively complicated in a way that feels complete to me